oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/tree/impl/AbstractTree.java

```diff
@@ -150,34 +151,59 @@ protected Iterable<String> getChildNames() {
 
     @Override
     public String toString() {
-        return toString(5);
+        return toJsonString(2);
     }
 
-    private String toString(int childNameCountLimit) {
-        StringBuilder sb = new StringBuilder();
-        sb.append(getPath()).append(": ");
-
-        sb.append('{');
-        for (PropertyState p : getProperties()) {
-            sb.append(' ').append(p).append(',');
-        }
-
-        Iterator<String> names = this.getChildNames().iterator();
-        int count = 0;
-        while (names.hasNext() && ++count <= childNameCountLimit) {
-            sb.append(' ').append(names.next()).append(" = { ... },");
+    /**
+     * Represents an OAK tree as a JSON object string
+     * When the representation is limited in depth, {"...":"..."} represents the limit.
+     * Note that, try-catch sentences are used to add "ERROR: "+e.getMessage() values in case of representation error.
+     * Note that, this is a recursive method.
+     * @param depth tree depth to represent. Use -1 for an unlimited depth.
+     * @return json object representation of the tree as a string
+     */
+    public String toJsonString(int depth) {
+        if (depth == 0) {
+            return quote("...");
         }
-
-        if (names.hasNext()) {
-            sb.append(" ...");
+        StringBuilder str = new StringBuilder();
+        str.append("{");
+        str.append(quote("_properties_") + ":{ ");
+        for (PropertyState ps : this.getProperties()) {
+            str.append(quote(ps.getName()) + ":");
+
+            if (ps.getType().isArray()) {
+                str.append("[ ");
+                for (int i = 0; i < ps.count(); i++) {
+                    try {
+                        str.append(JsopBuilder.encode(ps.getValue(Type.STRING, i)) + ",");
+                    } catch (Exception e) {
+                        str.append(quote("ERROR:" + JsopBuilder.encode(e.getMessage())) + ",");
+                    }
+                }
+                str.deleteCharAt(str.length() - 1); //removing the space or the ,
+                str.append("],");
+            } else {
+                try {
+                    str.append(JsopBuilder.encode(ps.getValue(Type.STRING)) + ",");
+                } catch (Exception e) {
+                    str.append(quote("ERROR:" + JsopBuilder.encode(e.getMessage())) + ",");
+                }
+            }
         }
-
-        if (sb.charAt(sb.length() - 1) == ',') {
-            sb.deleteCharAt(sb.length() - 1);
+        str.deleteCharAt(str.length() - 1);  //removing the space or the ,
```

Comment:
I would avoid that approach. Just start with an empty separator string, and set it to ", " after first use.

Reply:
Yeah! We had a discussion with @thomasmueller and @joerghoh about this. We thought that better merge it to a branch so more people can contribute before it is PR towards the trunk.